### PR TITLE
docs: Update Custom Resources documentation

### PR DESCRIPTION
**Pull request by jbyrne-redhat**
A preview of the rendered HTML is available for review [here](http://file.emea.redhat.com/~jbyrne/gh158/custom_resources.html).

### CODE REVIEW

**Comment by wtrocki** [comment]
Slightly confused
Were there problems with: https://github.com/redhat-developer/app-services-operator/pull/222?

**Comment by jbyrne-redhat** [comment]
> Slightly confused
> Were there problems with: #222?

Hey, @wtrocki. No, I'm going section by section. #222 was for the controllers documentation. This is for the CRs themselves.

File: docs/custom_resources.adoc [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
+
+[id="CloudServicesRequest"]
 == CloudServicesRequest
 
-This custom resource is populated with services available to the provided access token. These services are listed under the status object. This resource is used to get a list of available services. For instance, the id of a Kafka service is used by `KafkaConnection`.
+The `CloudServicesRequest` CR is used to get a list of available services. The controller for the CR updates the CR with services that are available to the provided access token. These services are listed under the `status` subresource. For example, the ID of an available Kafka service is used by the `KafkaConnection` CR.
 
 === Example
+
+An example of the `CloudServicesRequest` CR is shown below.
```
Comment (wtrocki) [inline docs/custom_resources.adoc:18]:
```suggestion
An example of the `CloudServicesRequest` Custom Resource
```

File: docs/custom_resources.adoc [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
+* `KafkaConnection`
+* `CloudServiceAccountRequest`
+
+All CRs are namespace-scoped and include Operator-managed conditions. Each CR is described in the sections that follow.
+
+[id="CloudServicesRequest"]
 == CloudServicesRequest
 
-This custom resource is populated with services available to the provided access token. These services are listed under the status object. This resource is used to get a list of available services. For instance, the id of a Kafka service is used by `KafkaConnection`.
+The `CloudServicesRequest` CR is used to get a list of available services. The controller for the CR updates the CR with services that are available to the provided access token. These services are listed under the `status` subresource. For example, the ID of an available Kafka service is used by the `KafkaConnection` CR.
```
Comment (wtrocki) [inline docs/custom_resources.adoc:14]:
```suggestion
The `CloudServicesRequest` CR is used to get a list of available RHOAS services. The controller for the CR updates the CR with services that are available to the provided access token. These services are listed under the `status` subresource. For example, the ID of an available Kafka service is used by the `KafkaConnection` CR.
```

**Comment by wtrocki** [comment]
> A preview of the rendered HTML is available for review here.

We can enable github actions for this quite easily or netlify if you wish. 
It will be less hassle. I have enabled ghpages from branch. 

We can use https://asciidoctor.org to publish docs for now. 
I think it will be good to replicate this across the fleet of repositories like CLI. Operator etc.

Let me know what you think and consider it done.

**Comment by wtrocki** [comment]
Done. Once this PR is merged website with all docs will be available here:
https://redhat-developer.github.io/app-services-operator/

Every change that lands to master will rebuild website automatically:
https://github.com/redhat-developer/app-services-operator/blob/main/.github/workflows/website.yml

If you happy we can replicate that preview feature for each repo later.

**Review by wtrocki — APPROVED** [review]

File: docs/custom_resources.adoc [hunk 3]
```diff
@@ -205,43 +232,53 @@ status:
...
+All conditions have a `status` value of `True`, `False`, or `Unknown`. When the Operator processes a CR, it first sets the `status` value of each condition to `Unknown`. Then, the Operator sets each condition to a value of `True` as the condition is checked. If a condition fails to check, the Operator sets the `status` values of the condition *and* the `Finished` condition to `False`. The Operator also updates the `reason` and `message` fields with more information.
 
-Each custom resource type has several conditions on their status subresource. Some conditions are shared by multiple types and some are type specific. All types have the status of `True`, `False`, and `Unknown`. When the operator begins running on the resource, it will set all conditions to `Unknown`. The operator then sets each condition to `True` as they are checked. If a condition fails to check, it and the `Finished` conditions will be set to `False` as well as have their `reason` and `message` fields set.  Additionally, processing is halted and subsequent conditions are not checked and they will remain `Unknown`. In order to resume processing, the errors must be corrected and a new custom resource submitted.
```
Comment (wtrocki) [inline docs/custom_resources.adoc:239]: I have feeling that we missing some info that was removed here.

Comment (jbyrne-redhat) [reply docs/custom_resources.adoc:239]: @wtrocki I think you mean info removed from the current version in the 'app-services-operator' repo, and *not* from the PR I opened? I can't see anything missing from my updated version when I compare against the current version.

**Comment by jbyrne-redhat** [comment]
> > A preview of the rendered HTML is available for review here.
> 
> We can enable github actions for this quite easily or netlify if you wish.
> It will be less hassle. I have enabled ghpages from branch.
> 
> We can use https://asciidoctor.org to publish docs for now.
> I think it will be good to replicate this across the fleet of repositories like CLI. Operator etc.
> 
> Let me know what you think and consider it done.



> Done. Once this PR is merged website with all docs will be available here:
> https://redhat-developer.github.io/app-services-operator/
> 
> Every change that lands to master will rebuild website automatically:
> https://github.com/redhat-developer/app-services-operator/blob/main/.github/workflows/website.yml
> 
> If you happy we can replicate that preview feature for each repo later.

Thanks, @wtrocki! FYI, it's not that hard for a CCS writer to generate a preview as they author. Atom has a package that uses Asciidoctor to generate a preview version. I think it's very useful to include a link to the preview in the PR so that reviewers can see it, *prior* to merging.

But I do think this is a really useful feature for non-CCS contributors. 

The one repo where I think we **wouldn't** want this feature is the `app-services-guides` repo. In that repo, we have a different build process to preview quick starts. For standard documentation, CCS writers can easily generate their own previews, as I described previously.

**Comment by wtrocki** [comment]
Yes. Guides are not part of it for sure. This is only for individual code repositories

File: docs/custom_resources.adoc [hunk 4]
```diff
@@ -205,43 +232,53 @@ status:
...
- * ServiceAccountCreated - When status is `True` then the service account was created successfully.
- * ServiceAccountSecretCreated - When status is `True` then the credentials secret was created successfully.
- * Finished - see `Finished Condition`
- * AcccesTokenSecretValid - see `AcccesTokenSecretValid Condition`
+== Conditions common to all RHOAS Custom Resources
 
+Each CR type managed by the RHOAS Operator has several conditions defined in the `status` subresource. Some conditions are shared by multiple types and some are type-specific.
 
-== Conditions Common to RHOAS Custom Resources
+All conditions have a `status` value of `True`, `False`, or `Unknown`. When the Operator processes a CR, it first sets the `status` value of each condition to `Unknown`. Then, the Operator sets each condition to a value of `True` as the condition is checked. If a condition fails to check, the Operator sets the `status` values of the condition *and* the `Finished` condition to `False`. The Operator also updates the `reason` and `message` fields with more information.
```
Comment (wtrocki) [inline docs/custom_resources.adoc:272]:
```suggestion
Each CR has condition field that follow the same standard. Conditions are represented as array of objects.
All condition objects have a `status` value of `True`, `False`, or `Unknown`. When the Operator processes a CR, it first sets the `status` value of each condition to `Unknown`. Then, the Operator sets each condition to a value of `True` as the condition is checked. If a condition fails to check, the Operator sets the `status` values of the condition *and* the `Finished` condition to `False`. The Operator also updates the `reason` and `message` fields with more information.
```

**Review by wtrocki — APPROVED** [review]

**Comment by wtrocki** [comment]
Good to merge. I have added small suggestion that for me was making things more clear.

File: docs/custom_resources.adoc (hunk 1 above)
Comment (jbyrne-redhat) [reply docs/custom_resources.adoc:18]: I think this change treats the text as a caption for the code block, which it isn't. If it's OK, I think it's better to leave as-is.

Comment (wtrocki) [reply docs/custom_resources.adoc:18]: Agreed.

**Pull request merged**